### DISCUSSION TITLE
Error Pointers for APIViews

### DISCUSSION THREAD

**Discussion by exiareinert-hpa** [django-json-api/django-rest-framework-json-api#1044] [Q&A]
Hello!

I have an API view which inherits from `rest_framework.views.APIView`, rather than `GenericAPIView`. Before (I believe) PR #986 was merged, formatting for required field errors used field names, e.g.:

```
{
    "errors": [
        {
            "detail": "This field is required.",
            "status": "400",
            "source": {
                "pointer": "/data/attributes/address"
            },
            "code": "required"
        }
    ]
}
```

Because pointers are now set only if the view inherits from `GenericAPIView`, required field errors are now much less informative:

```
{
    "errors": [
        {
            "detail": "This field is required.",
            "status": "400",
            "code": "required"
        }
    ]
}
```

It doesn't make sense for my API view to inherit from `GenericAPIView`, because it provides the result of a calculation rather than directly returning model data. Would it be possible to reinstate the previous behavior in cases like this?

**Comment by sliverc** [comment]
Thanks for your feedback.

Before the fix of #986 the error pointer base was hard coded and was only correct in some cases and in other it was not. And the only way which seemed plausible is to make calculation based on the `GenericAPIView` as this is the lowest view class which enforces setting of serializer class. And the serializer class is needed to determine the correct pointer.

This said we might change this check but for this it is important to know the different APIView use cases.

So could you paste your `APIView` here to see how you have configured it? Thanks.

**Reply by exiareinert-hpa** [reply]
The API view (code example below) does use a serializer. Perhaps an alternate solution could involve manually declaring a serializer attribute? 

```
class CalculationView(APIView):
    permission_classes = (IsAuthenticated,)

    @swagger_auto_schema(
        query_serializer=CalculationSerializer,
        responses=custom_view_to_response("get", "CalculationView"),
    )
    def get(self, request, format=None):
        param_serializer = CalculationSerializer(data=request.GET)
        param_serializer.is_valid(raise_exception=True)

        calculator = Calculate(param_serializer.data)
        calculator.init()
        data = calculator.call()

        return Response(data)
```

**Reply by sliverc** [reply]
Thanks for the example. As you do not set `resource_name` in your view I assume that the data returned by `calculator.call()` is already in JSON:API spec format. Is this correct?

I think in such a case where the `APIView` is manually building the result it should also manually format the errors. For this you can overwrite [handle_exception](https://www.django-rest-framework.org/api-guide/views/#handle_exceptionself-exc) of your view and use the utility `rest_framework_json_api.utils.format_error_object` with a fix pointer to attributes when the exception is a single `ValidationError` or otherwise call super. I haven't tested this but should work in theory.
